### PR TITLE
Skip Exporting Specific Records

### LINKED ISSUE #16177: During incidents as an Operator of a Zeebe cluster I need a way of skipping records on processing or exporting
**Is your feature request related to a problem? Please describe.**

We had no multiple occurrences in the past weeks were a single record was either preventing processing or exporting to continue.

* [Error in rejection serialization blocks exporting](https://github.com/camunda/zeebe/issues/15989)
* ["Faulty" DMN causes on rejection partition one to be stuck](https://github.com/camunda/zeebe/issues/14929)

All these clusters running into these errors were not recoverable until a new patch was released. This might have an enormous impact on business availability. We need a faster and better way to recover such clusters running into these kind of issues. 

In the cases above it would have helped if we could simply skip the problematic record. 

> [!Important]
>
> This of course needs to be decided on a case-by-case basis and needs to be done with care.

**Describe the solution you'd like**

I would like to be able to skip certain specific records for example via position or if necessary a chunk of records via key (?). I think this can be discussed what is best here.

**Describe alternatives you've considered**

Right now we have no way to do this, users have to wait for new patches to recover.

**Additional context**

This could be quite severe depending when this happens in production so prepare something like this would be beneficial.

I see this as a ZDP topic as it is on the ExporterDirector and StreamProcessor level.

Related incidents:

 * https://app.slack.com/client/T0PM0P1SA/C06F3FDRFHU
 * https://app.slack.com/client/T0PM0P1SA/C06E5CBS20G

### PR DESCRIPTION
**Pull request by rodrigo-lourenco-lopes**
## Description

This pr allows the confiration of record's positions to skip in the exporters. A record position in this list will be treated as a record to skip similar to the event filter (`ExporterDirector.skipRecord()`).

The record positions to skip are being filtered out at the level of the export director, which means that a record position that is configured, will be skipped in all other exporters. 

This configuration can be changed on the broker config in `zeebe.broker.exporting.skipRecords` or in the env var `ZEEBE_BROKER_EXPORTING_SKIPRECORDS`.

## Related issues

closes https://github.com/camunda/zeebe/issues/16177

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: dist/src/main/config/broker.standalone.yaml.template [hunk 1]
```diff
@@ -725,6 +725,17 @@
...
+      #
+      # Enable the exporter to skip record position
+      #
+      # These setting can also be overridden using the environment variables "ZEEBE_BROKER_EXPORTERS_[exporter name]_SKIPRECORDS"
+      # Allows to skip certain records by their position. This is useful for debugging or skipping a record that is
+      # preventing processing or exporting to continue.
+      # Record positions defined to skip in one of the exporters will be skipped in all exporters.
+      # The value is a comma-separated list of records ids to skip. Whitespace is ignored.
+      # skipRecords:
+
```
Comment (oleschoenburg) [inline dist/src/main/config/broker.standalone.yaml.template:738]: :x: This is unexpected behavior and a bit confusing on multiple levels.
I think we should just add the `skipRecords` configuration on another level, so not in `zeebe.broker.exporters` but something like `zeebe.broker.exporting`.

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java [hunk 2]
```diff
@@ -391,7 +394,9 @@ private void skipRecord(final LoggedEvent currentEvent) {
   private void readNextEvent() {
     if (shouldExport()) {
       final LoggedEvent currentEvent = logStreamReader.next();
-      if (eventFilter == null || eventFilter.applies(currentEvent)) {
+      if (eventFilter == null
+          || eventFilter.applies(currentEvent)
+              && !positionsToSkip.contains(currentEvent.getPosition())) {
```
Comment (oleschoenburg) [inline zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java:399]: :wrench: We should reuse the `SkipPositionsFilter` I've built for skipping positions during processing.

**Review by lenaschoenburg — CHANGES_REQUESTED** [review]

File: dist/src/main/config/broker.standalone.yaml.template [hunk 3]
```diff
@@ -682,6 +682,20 @@
         # This setting can also be overridden using the environment ZEEBE_BROKER_BACKPRESSURE_GRADIENT2_LONGWINDOW
         # longWindow: 600
 
+    # exporting:
+      # This section contains definitions common to all exporters.
+      #
+      # Skip records Exporter ------------
```
Comment (oleschoenburg) [inline dist/src/main/config/broker.standalone.yaml.template:688]: :x: This appears to be left-over. There is no "Skip records Exporter". The following lines should be indented one level less.

File: dist/src/main/config/broker.yaml.template [hunk 4]
```diff
@@ -575,6 +575,20 @@
         # This setting can also be overridden using the environment ZEEBE_BROKER_BACKPRESSURE_GRADIENT2_LONGWINDOW
         # longWindow: 600
 
+    # exporting:
+      # This section contains definitions common to all exporters.
+      #
+      # Skip records Exporter ------------
```
Comment (oleschoenburg) [inline dist/src/main/config/broker.yaml.template:581]: :x: Same as above

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java [hunk 5]
```diff
@@ -280,7 +282,7 @@ private void initContainers() throws Exception {
       container.configureExporter();
     }
 
-    eventFilter = createEventFilter(containers);
+    eventFilter = createEventFilter(containers).and(positionsToSkipFilter);
```
Comment (oleschoenburg) [inline zeebe/broker/src/main/java/io/camunda/zeebe/broker/exporter/stream/ExporterDirector.java:285]: :wrench: I'd flip the order to first use the position skip filter. That way the other filters don't even have to decode (possibly corrupted) events.

File: zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/configuration/ExporterCfg.java [hunk 6]
```diff
@@ -25,6 +26,8 @@ public final class ExporterCfg implements ConfigurationEntry {
   /** fully qualified class name pointing to the class implementing the exporter interface */
   private String className;
 
+  private Set<Long> skipRecords;
```
Comment (oleschoenburg) [inline zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/configuration/ExporterCfg.java:29]: :x: This needs to be removed here, the config moved.

**Review by lenaschoenburg — CHANGES_REQUESTED** [review]

**Review by lenaschoenburg — APPROVED** [review]
Nice :rocket:

**Pull request merged**